internal/provider/serverless_cluster_resource.go

```diff
@@ -0,0 +1,819 @@
+	Port        types.Int64  `tfsdk:"port"`
+	AWSEndpoint *awsEndpoint `tfsdk:"aws_endpoint"`
+	GCPEndpoint *gcpEndpoint `tfsdk:"gcp_endpoint"`
+}
+
+type awsEndpoint struct {
+	ServiceName      types.String `tfsdk:"service_name"`
+	AvailabilityZone types.List   `tfsdk:"availability_zone"`
+}
+
+type gcpEndpoint struct {
+	ServiceAttachmentName types.String `tfsdk:"service_attachment_name"`
+}
+
+type encryptionConfig struct {
+	EnhancedEncryptionEnabled types.Bool `tfsdk:"enhanced_encryption_enabled"`
+}
+
+type usage struct {
+	RequestUnit     types.String `tfsdk:"request_unit"`
+	RowBasedStorage types.Int64  `tfsdk:"row_based_storage"`
+	ColumnarStorage types.Int64  `tfsdk:"columnar_storage"`
+}
+
+type serverlessClusterResource struct {
+	provider *tidbcloudProvider
+}
+
+func NewServerlessClusterResource() resource.Resource {
+	return &serverlessClusterResource{}
+}
+
+func (r *serverlessClusterResource) Metadata(_ context.Context, req resource.MetadataRequest, resp *resource.MetadataResponse) {
+	resp.TypeName = req.ProviderTypeName + "_serverless_cluster"
+}
+
+func (r *serverlessClusterResource) Configure(_ context.Context, req resource.ConfigureRequest, resp *resource.ConfigureResponse) {
+	// Prevent panic if the provider has not been configured.
+	if req.ProviderData == nil {
+		return
+	}
+
+	var ok bool
+	if r.provider, ok = req.ProviderData.(*tidbcloudProvider); !ok {
+		resp.Diagnostics.AddError("Internal provider error",
+			fmt.Sprintf("Error in Configure: expected %T but got %T", tidbcloudProvider{}, req.ProviderData))
+	}
+}
+
+func (r *serverlessClusterResource) Schema(_ context.Context, _ resource.SchemaRequest, resp *resource.SchemaResponse) {
+	resp.Schema = schema.Schema{
+		MarkdownDescription: "serverless cluster resource",
+		Attributes: map[string]schema.Attribute{
+			"project_id": schema.StringAttribute{
+				MarkdownDescription: "The ID of the project. When not provided, the default project will be used.",
+				Optional:            true,
+			},
+			"cluster_id": schema.StringAttribute{
+				MarkdownDescription: "The ID of the cluster.",
+				Computed:            true,
+				PlanModifiers: []planmodifier.String{
+					stringplanmodifier.UseStateForUnknown(),
+				},
+			},
+			"display_name": schema.StringAttribute{
+				MarkdownDescription: "The display name of the cluster.",
+				Required:            true,
+				PlanModifiers: []planmodifier.String{
+					stringplanmodifier.UseStateForUnknown(),
+				},
+			},
+			"region": schema.SingleNestedAttribute{
+				MarkdownDescription: "The region of the cluster.",
+				Required:            true,
+				PlanModifiers: []planmodifier.Object{
+					objectplanmodifier.UseStateForUnknown(),
+				},
+				Attributes: map[string]schema.Attribute{
+					"name": schema.StringAttribute{
+						MarkdownDescription: "The unique name of the region. The format is `regions/{region-id}`.",
+						Required:            true,
+						PlanModifiers: []planmodifier.String{
```

Comment:
required schema does not need `UseStateForUnknown`. Please check the whole schema